executor/envd/api/routes.py

```
@@ -187,10 +217,225 @@ async def upload_file(
+
+                # Extract archive
+                with tarfile.open(tmp_path, "r:gz") as tar:
+                    # Get member names for tracking
+                    for member in tar.getnames():
+                        if member.startswith(".claude_session_id"):
+                            session_restored = True
+                        if member == ".git" or member.startswith(".git/"):
+                            git_restored = True
+
+                    # Extract all
+                    tar.extractall(path=str(workspace_path))
```

Ruff (0.15.6)

[error] Uses of `tarfile.extractall()`

(S202)